In [1]:
import pandas as pd

# Part 1: Load the dataset

In [2]:
categorial_url = 'chronic_kidney_disease_categorical.csv'
numerical_url = 'chronic_kidney_disease_numerical.csv'
categorial_set = pd.read_csv(categorial_url)
numerical_set = pd.read_csv(numerical_url)
print(categorial_set)
print(numerical_set)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

# chronic_kidney_disease_categorical
|column|feature/label|explanation|
|---|---|---|
|unique_id|feature|A unique id for each patient.|
|al|feature|The albumin level.|
|su|feature|The sugar level.|
|rbc|feature|Indicates whether red blood cells are normal or abnormal.|
|pc|feature|Indicates whether pus cells are normal or abnormal.|
|pcc|feature|Indicates whether pus cell clumps present or notpresent.|
|ba|feature|Indicates whether bacteria present or notpresent.|
|htn|feature|Indicates whether the patient has high blood pressure.|
|dm|feature|Indicates whether the patient has diabetes.|
|cad|feature|Indicates whether the patient has coronary artery disease.|
|appet|feature|Measure patient's appetite (good / poor).|
|pe|feature|Indicates whether the patient has pedal edema.|
|ane|feature|Indicates whether the patient has anemia.|
|Target|label|Indicates whether the patient has chronic kidney disease (ckd) or not (notckd).|

categorical feature examples
|column|
|---|
|pc|
|ba|
|appet|

numerical feature examples
|column|
|---|
|unique_id|
|al|
|su|

# chronic_kidney_disease_numerical
|column|feature/label|explanation|
|---|---|---|
|unique_id|feature|A unique id for each patient.|
|age|feature|The age of the patient (years).|
|bp|feature|The blood pressure value, measured in mm/Hg.|
|bgr|feature|The blood glucose random level, measured in mgs/dl.|
|bu|feature|The blood urea level, measured in mgs/dl.|
|sc|feature|The serum creatinine level, measured in mgs/dl.|
|sod|feature|The sodium level, measured in mEq/L.|
|pot|feature|The potassium level, measured in mEq/L.|
|hemo|feature|The hemoglobin level, measured in gms.|
|pcv|feature|The packed cell volume, expressed as in percentage.|
|wbcc|feature|The white blood cell count, measured in cells/cmm.|
|rbcc|feature|The blood cell count, measured in millions/cmm.|

categorical feature examples
|column|
|---|
|None|

numerical feature examples
|column|
|---|
|age|
|sod|
|wbcc|

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_dup = numerical_set.duplicated().sum()
print(numerical_dup)

# Drop these duplicate rows
numerical_cleaned = numerical_set.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_dup = categorial_set.duplicated().sum()
print(categorical_dup)

categorial_cleaned = categorial_set.drop_duplicates()


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, not all unique ids are present in both datasets because the length of the numerical set and categorial set are different. Since I am using inner merge, the rows that are missing from one of the datasets will be excluded from the combine table.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(numerical_cleaned, categorial_cleaned, on='unique_id', how='inner')
#Print the combined dataset
print(merged)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = merged.isnull().any(axis=1)
percentage_missing = (missing_rows.sum() / len(merged)) * 100

# Print %
print(percentage_missing)

# Drop these rows from the dataset
dataset_cleaned = merged.dropna()

# Print the Dataset
print(dataset_cleaned)

60.76923076923077
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn   

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
dataset_sorted = dataset_cleaned.sort_values(by='Target')
dataset_sorted.reset_index(drop=True, inplace=True)

# Print the dataset
print(dataset_sorted)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dataset = pd.get_dummies(dataset_sorted, columns=categorical_columns, drop_first=True)

# Print the dataset
print(dataset)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Step 1: Calculate the mean and standard deviation for numerical columns
# Create a mask to identify outliers
for column in numerical_columns:
    mean = dataset[column].mean()
    std_dev = dataset[column].std()
    
    # Define bounds for outliers
    upper_limit = mean + 3 * std_dev
    lower_limit = mean - 3 * std_dev
    
    # Step 2: Drop outliers from the dataset
    dataset = dataset[(dataset[column] >= lower_limit) & (dataset[column] <= upper_limit)]

# Print the dataset
print(dataset)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns:
    min_value = dataset[column].min()
    max_value = dataset[column].max()
    
    # Apply Min-Max normalization
    dataset[column] = (dataset[column] - min_value) / (max_value - min_value)

# Print the dataset
print(dataset)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
0    0.206897  0.059406  0.000000

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column does not provide any insights in terms of modeling and predictions, therefore it should be removed.

In [10]:
#Remove that column

# drop the 'unique_id' column
dataset.drop(columns=['unique_id'], inplace=True)  # drop the 'unique_id' column

# Print the dataset
print(dataset)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
0    0.059406  0.000000  0.653226  ...      F

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
csv_file_path = 'dataset.csv'
dataset.to_csv(csv_file_path, index=False)
print(f'DataFrame exported to {csv_file_path}')

# Export the dataframe to a new json file
json_file_path = 'dataset.json'
dataset.to_json(json_file_path, orient='records')
print(f'DataFrame exported to {json_file_path}')

DataFrame exported to dataset.csv
DataFrame exported to dataset.json


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

### Paste the markdown table here

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

|    age |    bp |       bgr |        bu |        sc |        sod |        pot |        hemo |      pcv |      wbcc |      rbcc |   al |   su |   rbc_normal |   pc_normal |   pcc_present |   ba_present |   htn_yes |   dm_yes |   cad_yes |   appet_poor |   pe_yes |   ane_yes |   Target_notckd |
|--------|-------|-----------|-----------|-----------|------------|------------|-------------|----------|-----------|-----------|------|------|---------------|-------------|---------------|--------------|------------|----------|----------|---------------|----------|-----------|-----------------|
| 0.2027 |  0.75 | 0.158798  | 0.196078  | 0.160494  | 0.166667   | 0.206897   | 0.059406    | 0.0      | 0.653226  | 0.257143  |  4.0 |  0.0 | True          | False       | True          | True         | False      | False    | False    | False         | False    | True      | False           |
| 0.9054 |  1.0  | 0.965665  | 0.522876  | 0.641975  | 0.666667   | 0.0        | 0.148515    | 0.225806 | 0.217742  | 0.057143  |  3.0 |  2.0 | False         | False       | True          | False        | True       | True     | True     | True          | False    | False     | False           |
| 0.7432 |  0.5  | 0.442060  | 0.901961  | 0.432099  | 0.5       | 0.793103   | 0.0         | 0.032258 | 0.395161  | 0.057143  |  2.0 |  0.0 | False         | False       | False         | False        | True       | True     | True     | True          | True     | True      | False           |
| 0.7297 |  0.75 | 0.150215  | 0.281046  | 0.234568  | 0.533333   | 0.793103   | 0.336634    | 0.322581 | 0.5       | 0.314286  |  2.0 |  0.0 | False         | True        | False         | False        | False      | False    | False    | False         | False    | False     | False           |
| 0.5405 |  0.0  | 0.399142  | 0.535948  | 0.358025  | 0.7       | 0.379310   | 0.207921    | 0.161290 | 0.830645  | 0.057143  |  1.0 |  0.0 | True          | True        | False         | False        | True       | True     | False    | False         | False    | False     | False           |
| 0.6757 |  0.75 | 0.253219  | 0.633987  | 0.777778  | 0.366667   | 0.655172   | 0.138614    | 0.193548 | 0.169355  | 0.114286  |  2.0 |  0.0 | False         | False       | False         | False        | True       | False    | False    | False         | False    | False     | False           |
| 0.7162 |  0.5  | 1.0       | 0.163399  | 0.111111  | 0.066667   | 0.206897   | 0.267327    | 0.387097 | 0.532258  | 0.371429  |  1.0 |  0.0 | False         | True        | False         | False        | False      | True     | False    | True          | False    | False     | False           |
| 0.7297 |  0.0  | 0.935622  | 0.169935  | 0.160494  | 0.333333   | 0.034483   | 0.019802    | 0.064516 | 0.879032  | 0.0       |  3.0 |  1.0 | True          | False       | True          | False        | True       | False    | False    | True          | False    | True      | False           |
| 0.0    |  0.0  | 0.103004  | 0.372549  | 0.074074  | 0.5       | 0.689655   | 0.217822    | 0.225806 | 1.0       | 0.514286  |  4.0 |  0.0 | False         | False       | False         | True         | False      | False    | False    | True          | False    | False     | False           |
| 0.8784 |  0.0  | 0.206009  | 0.751634  | 0.604938  | 0.533333   | 0.689655   | 0.366337    | 0.387097 | 0.879032  | 0.371429  |  4.0 |  0.0 | True          | True        | False         | False        | True       | True     | False    | True          | True     | False     | False           |
| 0.8514 |  0.25 | 0.618026  | 0.562092  | 0.728395  | 0.0       | 0.344828   | 0.168317    | 0.161290 | 0.580645  | 0.085714  |  4.0 |  3.0 | True          | False       | True          | True         | True       | True     | True     | False         | True     | True      | False           |
| 0.8784 |  0.25 | 0.639485  | 0.470588  | 0.395062  | 0.433333   | 0.517241   | 0.267327    | 0.322581 | 0.104839  | 0.171429  |  3.0 |  0.0 | True          | True        | True          | True         | True       | True     | True     | False         | False    | False     | False           |
| 0.7838 |  0.0  | 0.725322  | 0.313725  | 0.481482  | 0.566667   | 0.862069   | 0.178218    | 0.193548 | 0.258065  | 0.114286  |  4.0 |  1.0 | False         | False       | False         | True         | True       | True     | False    | True          | True     | False     | False           |
| 0.6622 |  0.5  | 0.618026  | 0.411765  | 0.432099  | 0.566667   | 0.689655   | 0.316832    | 0.354839 | 0.25      | 0.200000  |  3.0 |  1.0 | True          | False       | True          | True         | True       | True     | False    | False         | True     | False     | False           |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

One way to tackle this issue is to preprocess the data. Before passing data to the language model for conversion, ensure that the input data is clean, well-structured, and properly formatted. In this way, LLMs can better identify data and therefore reduce the chance to hallucinate.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Code to rename all the columns in the dataset

# Mapping dictionary for abbreviations to full names
mapping = {
    'age': 'Age (years)',
    'bp': 'Blood Pressure (mm/Hg)',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cells',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random (mg/dl)',
    'bu': 'Blood Urea (mg/dl)',
    'sc': 'Serum Creatinine (mg/dl)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count (cells/cmm)',
    'rbcc': 'Red Blood Cell Count (millions/cmm)',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'CKD or Not CKD'
}

# Step: Rename the DataFrame columns using the mapping
dataset.rename(columns=mapping, inplace=True)

# Display the updated DataFrame
print("Updated DataFrame with full names:")
print(dataset)

Updated DataFrame with full names:
     Age (years)  Blood Pressure (mm/Hg)  Blood Glucose Random (mg/dl)  \
0       0.202703                    0.75                      0.158798   
1       0.905405                    1.00                      0.965665   
3       0.743243                    0.50                      0.442060   
4       0.729730                    0.75                      0.150215   
6       0.540541                    0.00                      0.399142   
..           ...                     ...                           ...   
148     0.743243                    0.25                      0.214592   
149     0.418919                    0.00                      0.167382   
150     0.689189                    0.50                      0.210300   
151     0.472973                    0.25                      0.236052   
152     0.662162                    0.50                      0.145923   

     Blood Urea (mg/dl)  Serum Creatinine (mg/dl)  Sodium (mEq/L)  \
0      

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [13]:
# Convert SQL query to pandas query
result = dataset_sorted.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

This SQL query counts the number of ckd and notckd Targets in the dataset. 